### DISCUSSION TITLE
Calling a custom plugin method inside module

### DISCUSSION THREAD

**Discussion by joepagan** [craftcms/cms#9823] [Q&A] [answered]
[Asked in discord already](https://discord.com/channels/456442477667418113/456442867163070466/886927701523451915)

But basically I am trying to call a plugin service method in another module's init() function, I hope that I'm doing something stupid here?:

main Plugin file init() function:
```php
$this->setComponents([
          'foo' => Foo::class,
        ]);
```

Module file:
```php
use mynamespace\pluginhandle\PluginClass;
...
PluginClass::getInstance()->foo->bar()
```

Get the error:

```text
Attempt to read property "foo" on null
```

A helpful suggestion linked to [this discussion](https://github.com/craftcms/cms/discussions/9671) which suggests maybe plugins aren't available at the point that modules load? 
OP's solution here it to wrap this plugin method in an event `EVENT_AFTER_LOAD_PLUGINS`, though this is not really maintainable if you want to use it at multiple places throughout your module.

**Comment by brandonkelly** [comment, marked as answer]
> A helpful suggestion linked to [this discussion](https://github.com/craftcms/cms/discussions/9671) which suggests maybe plugins aren't available at the point that modules load?
> OP's solution here it to wrap this plugin method in an event `EVENT_AFTER_LOAD_PLUGINS`, though this is not really maintainable if you want to use it at multiple places throughout your module.

Yep that’s the way to go, as plugins are loaded _after_ modules. You will only need to worry about this trick for code in your module’s `init()` method (or called from it).

**Reply by joepagan** [reply]
Ah, thank you for letting me know! I had presumed it was going to be like that everywhere before even trying it somewhere else first!
